compiler/src/iree/compiler/Dialect/Flow/Transforms/InterchangeTransposeGenericOps.cpp

```diff
@@ -0,0 +1,109 @@
+// Copyright 2022 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+//===--------------- InterchangeTransposeGenericOps.cpp -------------------===//
+//
+// Interchange loops in generic ops to make the transpose happen on the outputs
+// instead of inputs.
+//
+//===----------------------------------------------------------------------===//
+
+#include "iree/compiler/Dialect/Flow/Transforms/FusionUtils.h"
+#include "iree/compiler/Dialect/Flow/Transforms/PassDetail.h"
+#include "iree/compiler/Dialect/Flow/Transforms/Passes.h"
+#include "mlir/Dialect/Linalg/IR/Linalg.h"
+#include "mlir/Dialect/Linalg/Transforms/Transforms.h"
+#include "mlir/Transforms/GreedyPatternRewriteDriver.h"
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+namespace {
+
+struct TransposeGenericOpPattern : public OpRewritePattern<linalg::GenericOp> {
+  using OpRewritePattern<linalg::GenericOp>::OpRewritePattern;
+  LogicalResult matchAndRewrite(linalg::GenericOp genericOp,
+                                PatternRewriter &rewriter) const override {
+    if (!genericOp.hasTensorSemantics()) return failure();
+
+    // indexing maps for linalg named ops
+    SmallVector<AffineMap> indexingMaps;
+
+    for (auto operand : genericOp.getInputOperands()) {
```

Comment:
The goal here is to group operations that can tile and fuse into one dispatch. So the indexing map is only relevant to look at when the producer is a LinalgOp. So I'd expect something like this

```
// Pattern needs to trigger only on  elementwise ops.
if (genericOp.getNumParallelLoops() != genericOp.getNumLoops()) return failure();

Optional<AffineMap> mapForInterchange = llvm::None;
for (auto opOperand : genericOp.getInputOperands()) {
  auto producer = opOperand->get().getDefiningOp<linalg::LinalgOp>();
  if (!producer) continue;
  
  auto indexingMap = genericOp.getTiedIndexingMap(opOperand);
  // This is already identity. Nothing to do.
  if (indexingMap.isIdentity()) return failure();
  // The map must be a permutation. If it is not then look for other operand.
  if (!indexingMap.isPermutation()) continue;

  if (!mapForInterchange) mapForInterchange = indexingMap;
}
if (!mapForInterchange) return failure();
auto interchange = llvm::to_vector(llvm::map_range(mapForInterchange->getResults(), [&](AffineExpr e) { return e.dyn_cast<AffineDimExpr>().getPosition(); }))

auto interchangedOp = intechangeGenericOp(rewriter, genericOp, intechange);
rewriter.replaceOp(genericOp, interchangedOp);
return success();
```

Reply:
Your code transposes only based on the first indexing map. When there are multiple linalg ops as operand, what do we expect? My code gives up because without performance estimation of the multiple cases, it is unclear which indexing map to choose.

Reply:
`interchangeGenericOp` actually does the op replacement.

Reply:
> Your code transposes only based on the first indexing map. When there are multiple linalg ops as operand, what do we expect? My code gives up because without performance estimation of the multiple cases, it is unclear which indexing map to choose.

The approach is, 
1) If there is already a linalg op that can be tiled and fused with the operation, then there is nothing to do.
2) If there is none, then picking any linalg operation is equivalent. The goal is to reduce the number of dispatches, so al options are "equally good" (at least for a first order of approximation). Its a heuristic, but is just the bare minimum heuristic.

Reply:
> `interchangeGenericOp` actually does the op replacement.

My bad. Missed that!